experimental/packages/otlp-transformer/src/trace/index.ts

```diff
@@ -0,0 +1,93 @@
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import type { InstrumentationLibrary } from '@opentelemetry/core';
+import type { Resource } from '@opentelemetry/resources';
+import type { ReadableSpan } from '@opentelemetry/sdk-trace-base';
+import { toAttributes } from '../common/internal';
+import { toSpan } from './internal';
+import { IExportTraceServiceRequest } from './types';
+
+export function createExportTraceServiceRequest(spans: ReadableSpan[]): IExportTraceServiceRequest | null {
+  if (spans.length === 0) {
+    return null;
+  }
+
+  return {
+    resourceSpans: groupByResource(spans).map(({ resource, resourceSpans, schemaUrl: resourceSchemaUrl }) => ({
+      resource: {
+        attributes: toAttributes(resource.attributes),
+        droppedAttributesCount: 0,
+      },
+      instrumentationLibrarySpans: resourceSpans.map(({ instrumentationLibrary, instrumentationLibrarySpans, schemaUrl: librarySchemaUrl }) => ({
+        instrumentationLibrary: {
+          name: instrumentationLibrary.name,
+          version: instrumentationLibrary.version,
+        },
+        spans: instrumentationLibrarySpans.map(toSpan),
+        schemaUrl: librarySchemaUrl,
+      })),
+      schemaUrl: resourceSchemaUrl,
+    }))
+  };
+}
+
+type ResourceSpans = {
+  resource: Resource,
+  resourceSpans: InstrumentationLibrarySpans[],
+  schemaUrl?: string,
+};
+
+type InstrumentationLibrarySpans = {
+  instrumentationLibrary: InstrumentationLibrary,
+  instrumentationLibrarySpans: ReadableSpan[],
+  schemaUrl?: string,
+};
+
+function groupByResource(SpanRecords: ReadableSpan[]): ResourceSpans[] {
+  const resourceMap: Map<Resource, Map<string, ReadableSpan[]>> = new Map();
+
+  for (const record of SpanRecords) {
+    let ilmMap = resourceMap.get(record.resource);
+
+    if (!ilmMap) {
+      ilmMap = new Map();
+      resourceMap.set(record.resource, ilmMap);
+    }
+
+    const instrumentationLibraryKey = JSON.stringify([record.instrumentationLibrary.name, record.instrumentationLibrary.version, record.instrumentationLibrary.schemaUrl]);
+    let records = ilmMap.get(instrumentationLibraryKey);
+
+    if (!records) {
+      records = [];
+      ilmMap.set(instrumentationLibraryKey, records);
+    }
+
+    records.push(record);
+  }
+
+  const out: ResourceSpans[] = [];
+
+  for (const [resource, ilmMap] of resourceMap.entries()) {
+    const resourceSpans: InstrumentationLibrarySpans[] = [];
+    for (const [instrumentationLibraryKey, instrumentationLibrarySpans] of ilmMap.entries()) {
+      const [name, version, schemaUrl] = JSON.parse(instrumentationLibraryKey);
```

Comment:
what do you think about storing this value in the Map on the previous iteration instead of `JSON.parse`ing it?
Something like:
```
const resourceMap: Map<Resource, Map<string, { spans: ReadableSpan[], instrumentationLibrary: IInstrumentationLibrary, schemaUrl: string | null}>> = new Map();
```

Reply:
FWIW, we have a similar map structure `HashMap` in sdk-metrics-base https://github.com/open-telemetry/opentelemetry-js/blob/main/experimental/packages/opentelemetry-sdk-metrics-base/src/state/HashMap.ts

Reply:
I just grabbed it from the first entry in the list instead